### PR TITLE
[mlir][Transforms] Dialect conversion: add `originalType` param to materializations

### PR DESCRIPTION
**Pull request by matthias-springer**
This commit adds an optional `originalType` parameter to target materialization functions. Without this parameter, target materializations are underspecified.

Note: `originalType` is only needed for target materializations. Source/argument materializations do not have it.

Consider the following example: Let's assume that a conversion pattern "P1" replaced an SSA value "v1" (type "t1") with "v2" (type "t2"). Then a different conversion pattern "P2" matches an op that has "v1" as an operand. Let's furthermore assume that "P2" determines that the legalized type of "t1" is "t3", which may be different from "t2". In this example, the target materialization callback will be invoked with: outputType = "t3", inputs = "v2", originalType = "t1". Note that the original type "t1" cannot be recovered from just "t3" and "v2"; that's why the `originalType` parameter is added.

This change is in preparation of merging the 1:1 and 1:N dialect conversion drivers. As part of that change, argument materializations will be removed (as they are no longer needed; they were just a workaround because of missing 1:N support in the dialect conversion). The new `originalType` parameter is needed when lowering MemRef to LLVM. During that lowering, MemRef function block arguments are replaced with the elements that make up a MemRef descriptor. The type converter is set up in such a way that the legalized type of a MemRef type is an `!llvm.struct` that represents the MemRef descriptor. When the bare pointer calling convention is enabled, the function block arguments consist of just an LLVM pointer. In such a case, a target materialization will be invoked to construct a MemRef descriptor (output type = `!llvm.struct<...>`) from just the bare pointer (inputs = `!llvm.ptr`). The original MemRef type is required to construct the MemRef descriptor, as static sizes/strides/offset cannot be inferred from just the bare pointer.

### CODE REVIEW

File: mlir/include/mlir/Transforms/DialectConversion.h [hunk 1]
```diff
@@ -303,20 +323,22 @@ class TypeConverter {
...
+                                 resultType, inputs, originalType);
   }
   Value materializeSourceConversion(OpBuilder &builder, Location loc,
-                                    Type resultType, ValueRange inputs) const {
+                                    Type resultType, ValueRange inputs,
+                                    Type originalType) const {
     return materializeConversion(sourceMaterializations, builder, loc,
-                                 resultType, inputs);
+                                 resultType, inputs, originalType);
   }
```
Comment (zero9178) [inline mlir/include/mlir/Transforms/DialectConversion.h:336]: If `originalType` should always be equal to `resultType`, I wouldn't add this as an extra parameter to these to reduce downstream churn and potential for errors

File: mlir/include/mlir/Transforms/DialectConversion.h [hunk 2]
```diff
@@ -170,21 +170,27 @@ class TypeConverter {
...
   /// converting (potentially multiple) block arguments that were the result of
   /// a signature conversion of a single block argument, to a single SSA value
   /// with the old block argument type.
+  ///
+  /// Note: The original type matches the result type `T` for argument
+  /// materializations.
   template <typename FnT, typename T = typename llvm::function_traits<
-                              std::decay_t<FnT>>::template arg_t<1>>
+                              std::decay_t<FnT>>::template arg_t<2>>
   void addArgumentMaterialization(FnT &&callback) {
```
Comment (zero9178) [inline mlir/include/mlir/Transforms/DialectConversion.h:194]: Could we nevertheless allow to pass callables to this and source materializations that do not have the `Type` parameter as fifth argument? 
I think this should greatly reduce churn downstream and also keep argument and source materializations cleaner conceptually. 

I believe this should be achievable by either usign `if constexpr` or an extra overload of `wrapMaterialization` by checking whether its callable using the 5 parameter variant or the 4 parameter variant.

**Review by zero9178 — COMMENTED** [review]
I see uses of `materializeTargetConversion` within normal conversion patterns where its non-obvious to me as to how that method should now be used.

With your example of "v2" replacing "v1", the adaptor of the conversion pattern would then receive "v2", is that right? 
Isn't t1 then also inaccessible to the conversion pattern? Or can the case you describe not happen during conversion pattern application OR should `materializeTargetConversion` not actually ever be called by conversion patterns?

**Comment by matthias-springer** [comment]
> I see uses of `materializeTargetConversion` within normal conversion patterns where its non-obvious to me as to how that method should now be used.

Ideally, patterns should not call this function at all. It's not really an API violation. I'm just saying that it should not be necessary. The uses that I've seen in upstream MLIR are workarounds around 1:N limitations and they are going to disappear soon. Worst case, users can always pass `originalType = Type()` and ignore this new parameter. (The dialect conversion itself does not require this parameter, only the target materializations may need it; but these are also written by the user.)

> With your example of "v2" replacing "v1", the adaptor of the conversion pattern would then receive "v2", is that right?

Almost. The adaptor receives "v2" only if the type of "v2" is valid according to the pattern's type converter. Otherwise, we will insert another target materialization from "v2" -> "t3" and that value is then passed to the adaptor. That's what's happening during the MemRef->LLVM lowering that I mentioned in the commit message.

The logic for this is in `ConversionPatternRewriterImpl::remapValues`. The implementation has access to the original types; the same is the case for the patterns themselves: they can always get the original types from `op` instead of `adaptor`. However, they may not be able to get original types for ops other than the matched op. (Because they may have to traverse "old" IR which is problematic in a dialect conversion.)

File: mlir/include/mlir/Transforms/DialectConversion.h (hunk 2 above)
Comment (matthias-springer) [reply mlir/include/mlir/Transforms/DialectConversion.h:194]: Yes, it should be possible to accept multiple lambda overloads. We do the same for `addConversion`. Let me give it a try.

File: mlir/include/mlir/Transforms/DialectConversion.h (hunk 1 above)
Comment (matthias-springer) [reply mlir/include/mlir/Transforms/DialectConversion.h]: Good idea, the PR became much smaller and no change is needed to existing code.

File: mlir/include/mlir/Transforms/DialectConversion.h [hunk 3]
```diff
@@ -151,15 +152,7 @@ class TypeConverter {
   ///       existing value are expected to be removed during conversion. If
   ///       `std::nullopt` is returned, the converter is allowed to try another
   ///       conversion function to perform the conversion.
-  ///   * std::optional<LogicalResult>(T, SmallVectorImpl<Type> &,
```
Comment (zero9178) [inline mlir/include/mlir/Transforms/DialectConversion.h:154]: Is this an unrelated change?

**Review by zero9178 — APPROVED** [review]
LGTM now, thank you!

> Ideally, patterns should not call this function at all. It's not really an API violation. I'm just saying that it should not be necessary. The uses that I've seen in upstream MLIR are workarounds around 1:N limitations and they are going to disappear soon.

Once your 1:N changes are merged, I think I'd be absolutely amazing to get rid of the API entirely then. I know at least from my own code and other downstream code that it is used rather defensively. But if it is really not needed then it'd help to conceptually simplify dialect conversion by not making these functions public.

File: mlir/include/mlir/Transforms/DialectConversion.h (hunk 3 above)
Comment (matthias-springer) [reply mlir/include/mlir/Transforms/DialectConversion.h:154]: I just noticed that this overload does not exist anymore.

**Review by ingomueller-net — APPROVED** [review]
LGTM! Interesting corner case -- glad you caught it. This makes sense to fix even without the merging. But I am looking forward to the merging anyways :)
